### PR TITLE
DOCKER-630: Allow user to pass locality hint/restriction in docker run/create (swarm-iness)

### PR DESCRIPTION
**Pull request by trentm**

### CODE REVIEW

File: lib/backends/sdc/affinity.js [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+var errors = require('../../errors');
+
+
+// ---- globals
+
+var FILTER_KEY_RE = /^[a-z_][a-z0-9\-_.]+$/i;
+// Note that this differs from the Swarm one in expr.go to fix some issues
+// and accomodate slight parsing differences.
+// JSSTYLED
+var FILTER_VALUE_RE = /^[-a-z0-9:_\s.*/()?+[\]\\^$|]+$/i;
```
Comment (joshwilsdon) [inline lib/backends/sdc/affinity.js]: Might be good to list which types of values this is supposed to allow. Regexes are not the most readable.

File: test/unit/affinity.test.js [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+    var o = obj;
+    var parts = lookup.split('.');
+    var s = [];
+    for (var i = 0; i < parts.length; i++) {
+        var part = parts[i];
+        if (part === 'tag') {
+            part = 'tags';  // VMAPI "?tag.foo=bar" looks up "tags.foo".
+        }
+        s.push(part);
+        if (!hasOwnProperty.call(o, part)) {
```
Comment (joshwilsdon) [inline test/unit/affinity.test.js]: To me:

```
if (!o.hasOwnProperty(part)) {
```

is much more readable.

File: tools/jsl.node.conf [hunk 3]
```diff
@@ -122,6 +122,7 @@
 +define setInterval
 +define setTimeout
 +define setImmediate
++define hasOwnProperty
```
Comment (joshwilsdon) [inline tools/jsl.node.conf]: This is unneeded if you use o.hasOwnProperty() as suggested at L53 of test/unit/affinity.test.js

File: lib/backends/sdc/affinity.js [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+            if (err && err.statusCode !== 404) {
+                cb(err);
+            } else if (!err && vm_ && vm_.docker
+                && ['destroyed', 'failed'].indexOf(vm_.state) === -1
+                && vm_.internal_metadata['docker:id'] === aff.value)
+            {
+                /*
+                 * If the 'docker:id' check fails, treat this like a 404.
+                 * This could happen if someone made up a docker_id that
+                 * started with the 32-chars for a real VM.
```
Comment (joshwilsdon) [inline lib/backends/sdc/affinity.js]: This block comment doesn't look like it's describing the code here. Does that belong in the else?

**Comment by trentm** [comment]
squash merged this branch to master.

**Pull request closed without merging**